app/import_functions.go

```diff
@@ -1179,140 +1227,171 @@ func (a *App) importDirectChannel(data *DirectChannelImportData, dryRun bool) *m
 	return nil
 }
 
-func (a *App) importDirectPost(data *DirectPostImportData, dryRun bool) *model.AppError {
-	var err *model.AppError
-	if err = validateDirectPostImportData(data, a.MaxPostSize()); err != nil {
-		return err
+func (a *App) importMultipleDirectPosts(data []*DirectPostImportData, dryRun bool) *model.AppError {
+	if len(data) == 0 {
+		return nil
+	}
+
+	for _, postData := range data {
+		if err := validateDirectPostImportData(postData, a.MaxPostSize()); err != nil {
+			return err
+		}
 	}
 
 	// If this is a Dry Run, do not continue any further.
 	if dryRun {
 		return nil
 	}
 
-	var userIds []string
-	for _, username := range *data.ChannelMembers {
-		var user *model.User
-		user, err = a.Srv.Store.User().GetByUsername(username)
-		if err != nil {
-			return model.NewAppError("BulkImport", "app.import.import_direct_post.channel_member_not_found.error", nil, err.Error(), http.StatusBadRequest)
-		}
-		userIds = append(userIds, user.Id)
+	users := make(map[string]*model.User)
+	postsWithData := []PostAndData{}
+	postsForCreateList := []*model.Post{}
+	postsForOverwriteList := []*model.Post{}
+	allUsers, err := a.Srv.Store.User().GetAll()
+	if err != nil {
+		return model.NewAppError("BulkImport", "app.import.import_direct_post.channel_member_not_found.error", nil, err.Error(), http.StatusBadRequest)
+	}
+	for _, user := range allUsers {
+		users[user.Username] = user
 	}
 
-	var channel *model.Channel
-	var ch *model.Channel
-	if len(userIds) == 2 {
-		ch, err = a.createDirectChannel(userIds[0], userIds[1])
-		if err != nil && err.Id != store.CHANNEL_EXISTS_ERROR {
-			return model.NewAppError("BulkImport", "app.import.import_direct_post.create_direct_channel.error", nil, err.Error(), http.StatusBadRequest)
+	for _, postData := range data {
+		var userIds []string
+		var ok bool
+		var err *model.AppError
+		for _, username := range *postData.ChannelMembers {
+			var user *model.User
+			if user, ok = users[username]; !ok {
+				return model.NewAppError("BulkImport", "app.import.import_direct_post.channel_member_not_found.error", nil, "", http.StatusBadRequest)
+			}
+			userIds = append(userIds, user.Id)
 		}
-		channel = ch
-	} else {
-		ch, err = a.createGroupChannel(userIds, userIds[0])
-		if err != nil && err.Id != store.CHANNEL_EXISTS_ERROR {
-			return model.NewAppError("BulkImport", "app.import.import_direct_post.create_group_channel.error", nil, err.Error(), http.StatusBadRequest)
+
+		var channel *model.Channel
+		var ch *model.Channel
+		if len(userIds) == 2 {
+			ch, err = a.GetOrCreateDirectChannel(userIds[0], userIds[1])
+			if err != nil && err.Id != store.CHANNEL_EXISTS_ERROR {
+				return model.NewAppError("BulkImport", "app.import.import_direct_post.create_direct_channel.error", nil, err.Error(), http.StatusBadRequest)
+			}
+			channel = ch
+		} else {
+			ch, err = a.createGroupChannel(userIds, userIds[0])
```

Comment:
Is it guaranteed that userIds will be 1 at minimum ?

Reply:
Yes, that is ensured by the fact that `postData.ChannelMembers` must be 2 or greater and that is ensured by the `validateDirectPostImportData` method.